File: сontrol-contest-01/C/java/Main.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        long a, b, c;

        Node(long a, long b, long c) {
            this.a = a;
            this.b = b;
            this.c = c;
        }
    }

    static int upperBound(long[] arr, double x) {
        int l = 0, r = arr.length;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (arr[mid] <= x) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return l;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder out = new StringBuilder();

        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            String[] nm = br.readLine().split(" ");
            int n = Integer.parseInt(nm[0]);
            int m = Integer.parseInt(nm[1]);

            long[] k = new long[n];
            for (int i = 0; i < n; i++) {
                k[i] = Long.parseLong(br.readLine());
            }
            Arrays.sort(k);

            Node[] p = new Node[m];
            for (int i = 0; i < m; i++) {
                String[] s = br.readLine().split(" ");
                long a = Long.parseLong(s[0]);
                long b = Long.parseLong(s[1]);
                long c = Long.parseLong(s[2]);
                p[i] = new Node(a, b, c);
            }

            for (int i = 0; i < m; i++) {
                double val = 2.0 * Math.sqrt((double) p[i].a * p[i].c);
                double kx_m = p[i].b - val;
                double kx_p = p[i].b + val;

                int pos = upperBound(k, kx_m);
                if (pos != k.length) {
                    long ans = k[pos];
                    if (ans < kx_p) {
                        out.append("YES\n").append(ans).append('\n');
                    } else {
                        out.append("NO\n");
                    }
                } else {
                    out.append("NO\n");
                }
            }

            out.append('\n');
        }

        System.out.print(out);
    }
}

File: сontrol-contest-01/C/cpp/main.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define vec vector

struct node {
    ll a, b, c;
};

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    int t = 1; cin >> t;
    while(t-->0){
        ll n, m; cin >> n >> m;
        vec<ll> k(n);
        for(int i = 0; i < n; i++){
            cin >> k[i];
        }
        sort(k.begin(), k.end());
        vec<node> p(m);
        for(int i = 0; i < m; i++){
            cin >> p[i].a >> p[i].b >> p[i].c;
        }
        for(int i = 0; i < m; i++){
            double long kx_m = p[i].b - (pow((4 * p[i].b * p[i].b - 4 * (p[i].b * p[i].b - 4 * p[i].a * p[i].c)),  0.5)) / 2;
            double long kx_p = p[i].b + (pow((4 * p[i].b * p[i].b - 4 * (p[i].b * p[i].b - 4 * p[i].a * p[i].c)),  0.5)) / 2;
            if (upper_bound(k.begin(), k.end(), kx_m) != k.end()){
                ll ans = *upper_bound(k.begin(), k.end(), kx_m);
                if (ans < kx_p) cout << "YES" << '\n' << ans << '\n';
                else cout << "NO" << '\n';
            }else {
                cout << "NO" << '\n';
            }
        }
        cout << '\n';
    }
}